src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -697,3 +719,76 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
         Ok(unsafe { rwh_06::DisplayHandle::borrow_raw(raw.into()) })
     }
 }
+
+#[derive(Debug)]
+struct PumpEventNotifier {
+    /// Whether we're in winit or not.
+    control: Arc<(Mutex<PumpEventNotifierAction>, Condvar)>,
+    /// Thread handle.
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Drop for PumpEventNotifier {
+    fn drop(&mut self) {
+        *self.control.0.lock().unwrap() = PumpEventNotifierAction::Exit;
+        self.control.1.notify_one();
+        if let Some(handle) = self.handle.take() {
+            let _ = handle.join();
+        }
+    }
+}
+
+impl PumpEventNotifier {
+    fn spawn(connection: Connection, awakener: Ping) -> Self {
+        // Start from the waiting state.
+        let control = Arc::new((Mutex::new(PumpEventNotifierAction::Pause), Condvar::new()));
+        let control_thread = Arc::clone(&control);
+
+        let _ = connection.prepare_read().unwrap();
+
+        let handle =
+            std::thread::Builder::new().name(String::from("pump_events mon")).spawn(move || {
+                let (lock, cvar) = &*control_thread;
+                'outer: loop {
+                    let mut wait = lock.lock().unwrap();
+                    loop {
+                        match *wait {
+                            PumpEventNotifierAction::Monitor => break,
+                            PumpEventNotifierAction::Exit => break 'outer,
+                            PumpEventNotifierAction::Pause => (),
+                        }
+
+                        wait = cvar.wait(wait).unwrap();
+                    }
+                    // Wake-up the main loop and put this one back to sleep.
+                    *wait = PumpEventNotifierAction::Pause;
+                    drop(wait);
+
+                    while let Some(read_guard) = connection.prepare_read() {
+                        let _ = connection.flush();
+                        let poll_fd = PollFd::from_borrowed_fd(connection.as_fd(), PollFlags::IN);
+                        let _ = rustix::event::poll(&mut [poll_fd], -1);
```

Comment:
This will cause the join to deadlock.

Reply:
Added pipe to wake-up the thread just in case.